fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/RequestedOrderingConstraint.java

```diff
@@ -34,14 +35,40 @@ public class RequestedOrderingConstraint implements PlannerConstraint<Set<Reques
 
     @Nonnull
     @Override
-    public Optional<Set<RequestedOrdering>> combine(@Nonnull final Set<RequestedOrdering> currentConstraint, @Nonnull final Set<RequestedOrdering> newConstraint) {
-        if (currentConstraint.containsAll(newConstraint)) {
+    public Optional<Set<RequestedOrdering>> combine(@Nonnull final Set<RequestedOrdering> currentConstraint,
+                                                    @Nonnull final Set<RequestedOrdering> newConstraint) {
+        final var newRequestedOrderings = Sets.newLinkedHashSet(newConstraint);
+        for (final var newRequestedOrdering : newConstraint) {
+            for (final var currentRequestedOrdering : currentConstraint) {
+                // try to figure out if current already subsumes new
+                if (newRequestedOrdering.isDistinct() != currentRequestedOrdering.isDistinct()) {
+                    continue;
+                }
+                final var newOrderingParts = newRequestedOrdering.getOrderingParts();
+                final var currentOrderingParts = currentRequestedOrdering.getOrderingParts();
+
+                if (!currentRequestedOrdering.isExhaustive() && newRequestedOrdering.isExhaustive()) {
+                    continue;
+                }
+                if (currentRequestedOrdering.isExhaustive() && newOrderingParts.size() >= currentOrderingParts.size()) {
+                    if (newOrderingParts.subList(0, currentOrderingParts.size()).equals(currentOrderingParts)) {
```

Comment:
It would probably make sense to wrap this in a method that returns `true` if a `RequestedOrdering` prefixes / subsumes another `RequestedOrdering` to make the intention clearer.

By the way, it looks like `combine` is not symmetric, right?
for example:
`combine( (Set.of('a', 'b'), exhaustive=true), (Set.of('a', 'b', 'c'), exhaustive=true)) != combine( (Set.of('a', 'b', 'c'), exhaustive=true), (Set.of('a', 'b'), exhaustive = true))` since the first `combine` would return an empty set, while the second `combine` would return `Set.of('a', 'b'), Set.of('a', 'b', 'c'), exhaustive = true` if I am not mistaken, is this expected?

Reply:
Yes, it is expected but I think it is symmetric. If you already have a request for exhaustively producing all possible orders for `a, b`, you don't have to honor anything with a prefix of `a, b, c`. If on the other hand you already have `a, b, c` and then get `a, b` you just continue with `a, b`. In the first case you return empty since you don't want the set of required orders to change (but mind you `a, b` is already in there) while in the second case you want to promote the set of requested orders to something more encompassing: you go from `a, b, c` to `a, b`.

Reply:
New issue to be opened.